attestation-service/verifier/src/eventlog/mod.rs

```diff
@@ -0,0 +1,186 @@
+// Copyright (c) 2024 Alibaba Cloud
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+mod hash;
+
+use std::str::FromStr;
+
+use anyhow::{anyhow, bail, Context, Result};
+use hash::HashAlgorithm;
+use serde_json::{Map, Value};
+use sha2::{digest::FixedOutput, Digest, Sha256, Sha384, Sha512};
+
+#[derive(Clone)]
+pub struct AAEvent {
+    pub domain: String,
+    pub operation: String,
+    pub content: String,
+}
+
+impl FromStr for AAEvent {
+    type Err = anyhow::Error;
+
+    fn from_str(input: &str) -> Result<Self> {
+        let input_trimed = input.trim_end();
+        let sections: Vec<&str> = input_trimed.split(' ').collect();
+        if sections.len() != 3 {
+            bail!("Illegal AA event entry format. Should be `<domain> <operation> <content>`");
+        }
+        Ok(Self {
+            domain: sections[0].into(),
+            operation: sections[1].into(),
+            content: sections[2].into(),
+        })
+    }
+}
+
+#[derive(Clone)]
+pub struct AAEventlog {
+    pub hash_algorithm: HashAlgorithm,
+    pub init_state: Vec<u8>,
+    pub events: Vec<AAEvent>,
+}
+
+impl FromStr for AAEventlog {
+    type Err = anyhow::Error;
+
+    fn from_str(input: &str) -> Result<Self> {
+        let event_lines = input.lines().collect::<Vec<&str>>();
+
+        let (initline, eventline) = event_lines
+            .split_first()
+            .ok_or(anyhow!("at least one line should be included in AAEL"))?;
+
+        // Init line looks like
+        // INIT sha256/0000000000000000000000000000000000000000000000000000000000000000
+        let init_line_items = initline.split_ascii_whitespace().collect::<Vec<&str>>();
+        if init_line_items.len() != 2 {
+            bail!("Illegal INIT event record.");
+        }
+
+        if init_line_items[0] != "INIT" {
+            bail!("INIT event should start with `INIT` key word");
+        }
+
+        let (hash_algorithm, init_state) = init_line_items[1].split_once('/').ok_or(anyhow!(
+            "INIT event should have `<sha-algorithm>/<init-PCR-value>` as content after `INIT`"
+        ))?;
+
+        let hash_algorithm = hash_algorithm
+            .try_into()
+            .context("parse Hash Algorithm in INIT entry")?;
+        let init_state = hex::decode(init_state).context("parse init state in INIT entry")?;
+
+        let events = eventline
+            .iter()
+            .map(|line| AAEvent::from_str(line))
+            .collect::<Result<Vec<AAEvent>>>()?;
```

Comment:
can be a little less intimidating when written as `let events: Vec<AAEvent> = eventline.iter.map(AAEvent::from_str).collect()?;`

Reply:
My compiler will warn for this.
`line` is `&&str` but `AAEvent::from_str` only accepts `&str`. Thus we need to explicitly write the calling routine rather than simply the name of function to leverage auto dereference.

Also, https://doc.rust-lang.org/rust-by-example/error/iter_result.html#fail-the-entire-operation-with-collect gives exampl. Once a `map` fails, the returned will be the error. It is very useful but looks usual.

Reply:
ah 2x yes:

it needs to be `into_iter()` to not get a double ref, I guess.
I didn't spot the result short circuit, I think turbofish is required  here, if you want to squeeze it it one line.

I would consider renaming the variable though, it should not be singular (eventline), it looks like we are iterating over characters of a str otherwise. maybe you can call the initial variable different:

```rust
        let all_lines: Vec<&str> = input.lines().collect();

        let (initline, eventlines) = event_lines
            .split_first()
            .ok_or(anyhow!("at least one line should be included in AAEL"))?;
```

Reply:
Yes, let me rename the variables.  `into_iter()` still not work ; |